### PR TITLE
consensus: add support for `Invalid` vote

### LINKED ISSUE #1176: consensus: Distinguish Invalid vote from Timeout
#### Summary

Currently, Provisioners cast the same vote (a signature over a empty hash) for both an invalid block and a timeout event (unknown candidate).

This prevents determining if the Provisioner received an invalid candidate or simply didn't receive any candidate.

#### Relevant Context

This change is necessary to decide if a block is invalid (a quorum of Invalid votes is needed) and possibly implement Slashing over the misbehaving Generator.

#### Possible Solution

 - Add a `valid` field to Reduction messages.
 - When casting a vote:
   - if the candidate is unknown (Selection timeout-ed) the `block_hash` field is empty only (`valid` can be ignored)
   - if the candidate is known, `block_hash` contains the block hash (`valid` can be set to `true` if the block is valid, and `false` otherwise)

Alternative solutions can be proposed

### PR DESCRIPTION
**Pull request by herr-seppia**
Resolves #1176

### CODE REVIEW

File: consensus/src/validation/step.rs [hunk 1]
```diff
@@ -73,39 +92,28 @@ impl<T: Operations + 'static> ValidationStep<T> {
...
         // Call Verify State Transition to make sure transactions set is valid
-        if let Err(err) = Self::call_vst(candidate, executor).await {
-            error!(event = "failed_vst_call", ?err);
-            return;
-        }
+        let vote = match Self::call_vst(candidate, executor).await {
+            Ok(_) => Vote::Valid(header.hash),
+            Err(err) => {
+                error!(event = "failed_vst_call", ?err);
+                Vote::Invalid(header.hash)
```
Comment (goshawk-3) [inline consensus/src/validation/step.rs:99]: Returning `Vote::Invalid(header.hash)`  must be  also done on a failure of verify_block_header (a few lines above), correct?

Comment (herr-seppia) [reply consensus/src/validation/step.rs:99]: Yes, you should be definitely right. 

Ps: Does it worth to add a byte to `Vote::Invalid` to specify the reason of the failure?
It would be useful to implement different kind of slash, but it would impact the vote signature aggregation

File: node-data/src/message.rs [hunk 2]
```diff
@@ -457,7 +421,7 @@ pub enum Payload {
     GetInv(payload::Inv),
     GetBlocks(payload::GetBlocks),
     GetData(payload::GetData),
-    CandidateResp(Box<payload::CandidateResp>),
+    CandidateResp(Box<payload::GetCandidateResp>),
```
Comment (fed-franz) [inline node-data/src/message.rs:424]: This should be `GetCandidateResp`

File: node-data/src/message.rs [hunk 3]
```diff
@@ -968,23 +1001,119 @@ impl<M: Clone> AsyncQueue<M> {
...
+    }
+    fn header_mut(&mut self) -> &mut ConsensusHeader {
+        &mut self.header
+    }
+}
+
+impl ConsensusMessage for Candidate {
+    fn signable(&self) -> Vec<u8> {
+        let mut signable = self.header.signable();
+        signable.extend_from_slice(&[ConsensusMsgType::Candidate as u8]);
```
Comment (fed-franz) [inline node-data/src/message.rs:1083]: why is this needed?

**Review by fed-franz — CHANGES_REQUESTED** [review]

File: consensus/src/validation/step.rs (hunk 1 above)
Comment (goshawk-3) [reply consensus/src/validation/step.rs:99]: An extra flag is out of the scope of this PR, not needed for now IMO.

Comment (fed-franz) [reply consensus/src/validation/step.rs:99]: As you say, it would impact aggregation...I'm not sure it's worth it. If the block is deemed invalid, we can always check it ourselves and decide accordingly.

File: consensus/src/step_votes_reg.rs [hunk 4]
```diff
@@ -110,18 +110,18 @@ impl IterationCerts {
         }
     }
 
-    fn for_hash(&mut self, hash: [u8; 32]) -> Option<&mut CertificateInfo> {
-        if hash == [0u8; 32] {
+    fn for_vote(&mut self, vote: &Vote) -> Option<&mut CertificateInfo> {
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/step_votes_reg.rs:113]: You're right
It's the time to support multiple votes there too!

File: node/src/databroker.rs [hunk 5]
```diff
@@ -362,7 +362,7 @@ impl DataBrokerSrv {
             return Err(anyhow::anyhow!("no items to fetch"));
         }
 
-        Ok(Message::new_get_data(inv))
+        Ok(Message::new_inv(inv))
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply node/src/databroker.rs:365]: If so, we should change the Topic in "Inv" instead of "GetData". I just changed to follow the convention used where the new refers to the topic of the message

File: consensus/src/validation/handler.rs [hunk 6]
```diff
@@ -121,18 +118,18 @@ impl MsgHandler<Message> for ValidationHandler {
 
             if quorum_reached {
                 // if the votes converged for an empty hash we invoke halt
-                if hash == [0u8; 32] {
+                if p.vote == Vote::NoCandidate {
                     tracing::warn!(
                         "votes converged for an empty hash (timeout)"
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/validation/handler.rs:123]: Right, going to change it to support Quorum::Invalid too

File: node-data/src/message.rs (hunk 2 above)
Comment (herr-seppia) [reply node-data/src/message.rs:424]: True

File: node-data/src/message.rs (hunk 3 above)
Comment (herr-seppia) [reply node-data/src/message.rs:1083]: Every consensus message is verified at the beginning of the phase. I agree with you that this can be useless, since the block is already signed, but I would investigate a bit more before remove it

File: consensus/src/ratification/step.rs [hunk 7]
```diff
@@ -69,6 +55,30 @@ impl<T: Operations + 'static, DB: Database> RatificationStep<T, DB> {
     }
 }
 
+pub fn build_ratification_payload(
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/ratification/step.rs:58]: `message::payload::Ratification` now includes a `ConsensusHeader`.
This, like `build_validation_payload`, is a utility to properly create a fully signed `message::payload::Ratification` (that now implements `ConsensusMessage`)

File: consensus/src/step_votes_reg.rs [hunk 8]
```diff
@@ -172,23 +172,24 @@ impl CertInfoRegistry {
...
-
-        let payload = payload::Quorum {
-            signature,
+        let mut payload = payload::Quorum {
+            header,
+            vote: result.vote,
             validation: result.cert.validation,
             ratification: result.cert.ratification,
         };
+        payload.sign(&ru.secret_key, ru.pubkey_bls.inner());
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/step_votes_reg.rs:190]: I added just to be compliant with current `ConsensusMessage::verify_signature`.

Next PR will introduce custom verify_signature for both Candidate and Quorum

File: node-data/src/message.rs [hunk 9]
```diff
@@ -275,34 +226,43 @@ impl Message {
     /// Creates topics.GetData  message
     pub fn new_get_data(p: payload::Inv) -> Message {
         Self {
-            header: Header::new(Topics::GetData),
+            topic: Topics::GetData,
             payload: Payload::GetInv(p),
```
[the remark this answers is not in the document]
Comment (fed-franz) [reply node-data/src/message.rs:230]: I know it's not part of this PR, but it seems that's the wrong payload to set (it's not just about naming). So I pointed it out. I guess it can be addressed in a different PR.

File: node-data/src/message.rs (hunk 3 above)
Comment (fed-franz) [reply node-data/src/message.rs:1083]: Agreed

File: node-data/src/message.rs [hunk 10]
```diff
@@ -411,35 +397,13 @@ impl Header {
...
-        pk: &dusk_bls12_381_sign::PublicKey,
-    ) -> [u8; 48] {
-        let msg = marshal_signable_vote(
-            self.round,
-            self.get_step(),
-            &self.block_hash,
-        );
+    pub fn signable(&self) -> Vec<u8> {
+        let mut buf = vec![];
+        buf.extend_from_slice(&self.round.to_le_bytes());
```
Comment (fed-franz) [inline node-data/src/message.rs:402]: Does signing `round` have any function?

File: node-data/src/message.rs [hunk 11]
```diff
@@ -311,45 +271,75 @@ impl Message {
...
-        self.header.topic
+        self.topic
     }
 }
 
 #[derive(Default, Clone, PartialEq, Eq)]
-pub struct Header {
-    pub topic: Topics,
+#[cfg_attr(any(feature = "faker", test), derive(fake::Dummy))]
+pub struct ConsensusHeader {
```
[the remark this answers is not in the document]
Comment (fed-franz) [reply node-data/src/message.rs:287]: Also, I'm not sure about the need for `msg_type` since we already have that information from the Payload type. Why are we adding this?

File: consensus/src/msg_handler.rs [hunk 12]
```diff
@@ -8,30 +8,25 @@ use crate::commons::{ConsensusError, RoundUpdate};
...
-use node_data::ledger::to_str;
-use node_data::message::{Message, MessageTrait, Status};
+use node_data::message::{Message, Status};
 use node_data::StepName;
-use std::fmt::Debug;
 use tracing::{debug, trace};
 
 /// Indicates whether an output value is available for current step execution
-/// (Step is Ready) or needs to collect data (Step is Pending)
-pub enum HandleMsgOutput {
```
Comment (goshawk-3) [inline consensus/src/msg_handler.rs:19]: I disagree with this change.  I chose Ready/Pending to reflect a pattern from Future task which is more descriptive to the  step execution. https://doc.rust-lang.org/std/task/enum.Poll.html

Comment (herr-seppia) [reply consensus/src/msg_handler.rs:19]: Ok, but I see that the inner message of the Pending enum is never used.

Do you agree if I rollback and remove that inner message?

Comment (goshawk-3) [reply consensus/src/msg_handler.rs:19]: Of course, if the inner of Pending is unused, we should remove it. In that way, it duplicates the Poll enum/pattern even more.

File: consensus/src/commons.rs [hunk 13]
```diff
@@ -146,22 +151,23 @@ impl QuorumMsgSender {
     /// Sends an quorum (internally) to the quorum loop.
     pub(crate) async fn send(&self, msg: Message) -> bool {
         if let Payload::Quorum(q) = &msg.payload {
-            if q.signature == [0u8; 48]
+            if q.header.signature.is_zeroed()
                 || q.validation.is_empty()
                 || q.ratification.is_empty()
-                || msg.header.block_hash == [0; 32]
+                || q.vote == Vote::NoCandidate
+            //TODO: Not sure about `q.vote == Vote::NoCandidate`
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/commons.rs:158]: Adding a comment to fix it while implementing https://github.com/dusk-network/rusk/issues/1268

File: consensus/src/step_votes_reg.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/step_votes_reg.rs]: Removed the signature of the Quorum Message

File: node/src/chain.rs [hunk 14]
```diff
@@ -20,6 +20,7 @@ use crate::{LongLivedService, Message};
 use anyhow::Result;
 use async_trait::async_trait;
 use dusk_consensus::commons::ConsensusError;
+pub use header_validation::verify_block_cert;
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply node/src/chain.rs:23]: It is required for the node bench

File: node-data/src/ledger.rs [hunk 15]
```diff
@@ -421,10 +423,18 @@ pub mod faker {
         }
     }
 
+    impl<T> Dummy<T> for bls::PublicKey {
+        fn dummy_with_rng<R: Rng + ?Sized>(_config: &T, rng: &mut R) -> Self {
+            let rand_val = rng.gen();
+            // let mut bls_key = [0u8; 96];
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply node-data/src/ledger.rs]: Right

File: node-data/src/message.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (herr-seppia) [reply node-data/src/message.rs:287]: > I'm not sure about the need for msg_type

It's just for derive properly the step from the header

File: node-data/src/message.rs (hunk 10 above)
Comment (herr-seppia) [reply node-data/src/message.rs:396]: We rely on the "round" when detecting out of sync, so for now I would leave it in the signed fields

File: node-data/src/message.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (herr-seppia) [reply node-data/src/message.rs:230]: @goshawk-3 what do you think about it?

[the remark this answers is not in the document]
Comment (goshawk-3) [reply node-data/src/message.rs:231]: > shouldn't this be `Payload::GetData` ?

This replicates bitcoin messages. 
"The format and maximum size limitations of the [“getdata” message](https://developer.bitcoin.org/reference/p2p_networking.html#getdata) are identical to the [“inv” message](https://developer.bitcoin.org/reference/p2p_networking.html#inv); only the message header differs."

src: https://developer.bitcoin.org/reference/p2p_networking.html#getdata

[the remark this answers is not in the document]
Comment (herr-seppia) [reply node-data/src/message.rs:242]: Thanks, now it's clear to me. I removed those changes

File: node-data/src/message.rs (hunk 2 above)
Comment (fed-franz) [reply node-data/src/message.rs:374]: This Payload name should also be `GetCandidateResp`

File: node/src/chain/header_validation.rs [hunk 16]
```diff
@@ -191,34 +193,35 @@ impl<'a, DB: database::DB> Validator<'a, DB> {
...
 ) -> anyhow::Result<(QuorumResult, QuorumResult)> {
     let committee = RwLock::new(CommitteeSet::new(curr_eligible_provisioners));
-
-    let hdr = node_data::message::Header {
-        topic: node_data::message::Topics::Unknown,
-        pubkey_bls: node_data::bls::PublicKey::default(),
-        round: height,
-        iteration,
-        block_hash,
+    let vote = match block_hash == [0; 32] {
```
Comment (fed-franz) [inline node/src/chain/header_validation.rs:205]: For the next PR `vote` should be passed as argument or even included in the certificate.

File: consensus/src/step_votes_reg.rs [hunk 17]
```diff
@@ -96,35 +103,28 @@ pub type SafeCertificateInfoRegistry = Arc<Mutex<CertInfoRegistry>>;
 
 #[derive(Clone)]
 struct IterationCerts {
-    valid: Option<CertificateInfo>,
-    nil: CertificateInfo,
+    votes: HashMap<Vote, CertificateInfo>,
+    no_candidate: CertificateInfo,
```
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:107]: there's no point in having `no_candidate` now that we have the hashmap

File: consensus/src/msg_handler.rs [hunk 18]
```diff
@@ -8,41 +8,41 @@ use crate::commons::{ConsensusError, RoundUpdate};
...
         &self,
-        msg: &T,
+        msg: &Message,
         ru: &RoundUpdate,
         iteration: u8,
         step: StepName,
         committee: &Committee,
         round_committees: &RoundCommittees,
     ) -> Result<(), ConsensusError> {
+        let from =
```
Comment (fed-franz) [inline consensus/src/msg_handler.rs:40]: I'd call this `signer` to avoid confusion with the sender of the message (our peer).
I'd probably rename also `get_pubkey_bls` to `get_signer`.
We can add these to the tech debt if you prefer

File: consensus/src/execution_ctx.rs [hunk 19]
```diff
@@ -291,37 +285,38 @@ impl<'a, DB: Database, T: Operations + 'static> ExecutionCtx<'a, DB, T> {
...
+                    msg.get_step(),
+                    msg,
+                );
+
+                return None;
+            }
+            Err(ConsensusError::PastEvent) => {
+                return self.process_past_events(msg).await;
+            }
+            Err(ConsensusError::InvalidValidation(QuorumType::NoQuorum)) => {
```
Comment (fed-franz) [inline consensus/src/execution_ctx.rs:305]: This case should be removed by the next PR, since NoQuorum is indeed a valid result for Validation. Maybe add a reminder

File: consensus/src/validation/handler.rs [hunk 20]
```diff
@@ -121,18 +122,19 @@ impl MsgHandler<Message> for ValidationHandler {
 
             if quorum_reached {
                 // if the votes converged for an empty hash we invoke halt
```
Comment (fed-franz) [inline consensus/src/validation/handler.rs:124]: this comment is obsolete

File: consensus/src/ratification/step.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (fed-franz) [reply consensus/src/ratification/step.rs:57]: I think the confusion stems from calling `ConsensusHeader` a header, while it's indeed something else. I'm gonna file an issue to rename it

**Review by fed-franz — APPROVED** [review]

**Review by fed-franz — APPROVED** [review]
Few minor comments. The rest LGTM
Good job

File: consensus/src/execution_ctx.rs (hunk 19 above)
Comment (herr-seppia) [reply consensus/src/execution_ctx.rs:305]: Indeed, I just added the specific case to log it as warn instead of Error

```
# before
{"timestamp":"2024-01-24T16:33:38.394082Z","level":"ERROR","message":"phase handler err: InvalidValidation(no_quorum)","target":"dusk_consensus::execution_ctx","spans":[{"iter":0,"name":"Ratification","pk":"pFPEcfxidLvwmFKR","round":17,"name":"main"}]}

# after
{"timestamp":"2024-01-29T10:12:36.322397Z","level":"WARN","event":"No quorum reached","iter":0,"target":"dusk_consensus::execution_ctx","spans":[{"iter":0,"name":"Ratification","pk":"pFPEcfxidLvwmFKR","round":1,"name":"main"}]}
```

File: consensus/src/step_votes_reg.rs (hunk 17 above)
Comment (herr-seppia) [reply consensus/src/step_votes_reg.rs:107]: Updated

File: consensus/src/msg_handler.rs (hunk 18 above)
Comment (herr-seppia) [reply consensus/src/msg_handler.rs:40]: Updated

File: consensus/src/validation/handler.rs (hunk 20 above)
Comment (herr-seppia) [reply consensus/src/validation/handler.rs:124]: Removed

**Review by fed-franz — APPROVED** [review]
